### ISSUE TITLE
Support custom bias / custom vocabulary for Speech to Text

### ISSUE THREAD

**Issue by manzke** [intrafind/ihub-apps#2244]
Support custom bias / custom vocabulary for Speec to Text

https://share.google/aimode/LM6LYBTL5WNYUaHIK

"Um die Fehlerquote bei Fachbegriffen, Eigennamen oder spezifischem Fachvokabular in Mistral Voxtral (z. B. Voxtral-Mini-4B-Realtime) zu minimieren, wird Context Biasing (auch Custom Vocabulary genannt) genutzt.Da vLLM die native Streaming-Inferenz-Engine für Voxtral ist, lässt sich dieses Feature direkt über Logits-Prozessoren oder die standardisierte OpenAI-kompatible API steuern.[1](https://www.google.com/goto?url=CAESUgHrOzAVTtPiJF5wOIlc2YJU2HrJXrGwJzkhvLEKgauvc8YybqFHhGCIbqEJUUnrLUGtjVpjBx2UZIy3ePYRsndgOOtx5YPsuU0OEA_X9t7DquM). Über die OpenAI-kompatible API (vLLM Server)Wenn Sie vLLM als Server ausführen (vllm serve mistralai/Voxtral-Mini-4B-Realtime-[2](https://www.google.com/goto?url=CAESVgHrOzAV5shKbC7iFxIcSCfI_wvNQXPAZjjWSVmOWV4H_qsspumja50soce4QPDIKrdcYg6SsUL6s6JgWoS1UsVtKXDbNqcQoF7dca3PFqBuUBHhFkFc)602), können Sie das Context Biasing bei Audio-Anfragen über den Parameter extra_body oder direkt über logit_bias steuern. Mistral nutzt intern ein standardisiertes JSON-Format für vLLM-Erweiterungen.Beispiel-Request (Python):pythonimport requests

# Ihre spezifischen Begriffe, die das Modell bevorzugen soll
custom_vocabulary = ["vLLM", "Voxtral", "Kubernetes", "PyTorch", "AI-Agent"]

payload = {
    "model": "mistralai/Voxtral-Mini-4B-Realtime-2602",
    "extra_body": {
        "context_biasing": {
            "words": custom_vocabulary,
            "bias_score": 5.0  # Erhöht die Wahrscheinlichkeit dieser Token-Sequenzen
        }
    }
}

# Senden des Audio-Inhalts an den vLLM /v[1](https://www.google.com/goto?url=CAESYwHrOzAVQpVgQIrnYjiV4ltfMZMVegxXOOEUuK6RDSAuBwYd96Zw72_uFLCqZLFMqAk62BdKah5JXeS5mDKc3D50UbmNMH9g6NCAXV8fRilzrRaHpGq-mXei6Y7Dr67Wl1xK5Q)/audio/transcriptions Endpunkt
# (Je nach vLLM-Version wird dies als multipart/form-data mit 'extra_body' übergeben)
Verwende Code mit Vorsicht.[2](https://www.google.com/goto?url=CAESTgHrOzAVuWP_hreO42Ms2x7xDSEeXWXydpcdUIPhjVq6LdTTiwP5pTzYt79IJM4CD1wfECGtrfMNGEeVVi5KXjnn7WkfhsAkROM4YxNN6g). Logit-Bias Methode (Token-Ebene)Falls die native context_biasing-Abstraktion in Ihrer vLLM-Pipeline noch manuell verdrahtet werden muss, nutzt man klassischen Logit Biasing. Hierbei erhöht man mathematisch die Wahrscheinlichkeit der Token, aus denen Ihr Custom Vocabulary besteht.pythonfrom vllm import LLM, SamplingParams

llm = LLM(model="mistralai/Voxtral-Mini-4B-Realtime-2602", tokenizer="mistralai/Voxtral-Mini-4B-Realtime-2602")
tokenizer = llm.get_tokenizer()

# Wörter in Token-IDs umwandeln
custom_words = ["Voxtral", "Inferenz"]
logit_bias = {}

for word in custom_words:
    token_ids = tokenizer.encode(word, add_special_tokens=False)
    for tid in token_ids:
        logit_bias[tid] = 4.5  # Ein Wert zwischen 1 und 5 reicht meist aus

# Sampling-Parameter mit Bias an vLLM übergeben
sampling_params = SamplingParams(
    temperature=0.0,
    logit_bias=logit_bias
)

# Audio-Inferenz starten
# outputs = llm.generate(inputs={"audio": audio_data}, sampling_params=sampling_params)
Verwende Code mit Vorsicht.Best Practices für Context Biasing bei VoxtralGewichtung balancieren: Wählen Sie den Bias-Score nicht zu hoch (Richtwert: 2.0 bis 5.0). Ein zu hoher Wert führt dazu, dass das Modell die Wörter halluziniert, selbst wenn sie nicht gesprochen wurden.Präfix-Caching nutzen: Da vLLM automatisches Prefix-Caching (--enable-prefix-caching) unterstützt, bleibt die Performance hoch, wenn Sie bei jedem Streaming-Request dieselbe Liste an Custom-Vokabular mitsenden.Groß-/Kleinschreibung beachten: Tokenizer reagieren empfindlich auf Leerzeichen und Fallstricke der Schreibweise. Fügen Sie wichtige Wörter idealerweise sowohl mit als auch ohne führendes Leerzeichen hinzu (z. B. " Voxtral" und "Voxtral").Möchten Sie Context Biasing für eine Echtzeit-Streaming-Anwendung (WebSockets) aufsetzen oder für Batch-Transkriptionen (Dateien) nutzen? Teilen Sie mir gerne mit, in welchem Framework (z. B. LiveKit, FastAPI oder direkt Python) Ihr Code läuft."
